llama/src/main/java/com/cloudera/llama/am/yarn/YarnRMConnector.java

```diff
@@ -806,10 +815,51 @@ public void onContainersAllocated(List<Container> containers) {
           LOG.trace("Reservation resource '{}' removed from YARN", resource);
 
           queue(new ContainerHandler(ugi, resource, container, Action.START));
+
+          /*Remove the granted request from anyLocationResourceIdToRequestMap if it is there*/
+          anyLocationResourceIdToRequestMap.remove(resource.getResourceId());
         }
       } else {
-        LOG.error("No matching request for {}. Releasing the container.",
+        LOG.debug("No strong request match for {}. Adding to the list of unclaimed containers.",
             container);
+        unclaimedContainers.add(container);
+      }
+    }
+    /*Matching YARN resources against requests relaxing locality*/
```

Comment:
After moving the duplicate code into the separate private method, onContainersAllocated method looks quite concise and readable. Please let me know if you still would like me to move weak match logic into separate method also.